pkg/blockbuilder/blockbuilder.go

```diff
@@ -427,39 +483,77 @@ func (b *BlockBuilder) blockUploaderForUser(ctx context.Context, userID string)
 	}
 }
 
-func (b *BlockBuilder) finalizePartition(ctx context.Context, uncommittedRec, checkpointRec *kgo.Record) error {
-	// If there is an uncommitted record, rewind the client to the record's offset to consume it on the next cycle.
-	if uncommittedRec != nil {
-		part := uncommittedRec.Partition
-		b.kafkaClient.SetOffsets(map[string]map[int32]kgo.EpochOffset{
-			b.cfg.Kafka.Topic: {
-				part: {
-					Epoch:  uncommittedRec.LeaderEpoch,
-					Offset: uncommittedRec.Offset - 1,
-				},
-			},
-		})
+func (b *BlockBuilder) finalizePartition(ctx context.Context, commitRec, lastRec *kgo.Record, blockEnd int64) error {
+	if commitRec == nil {
+		return nil
 	}
+	// Rewind the offset to the commit record so that when the partition is read again by this
+	// block builder, it starts at the commit point.
+	b.kafkaClient.SetOffsets(map[string]map[int32]kgo.EpochOffset{
+		b.cfg.Kafka.Topic: {
+			commitRec.Partition: {
+				Epoch:  commitRec.LeaderEpoch,
+				Offset: commitRec.Offset,
+			},
+		},
+	})
 
-	if checkpointRec != nil {
-		// TODO(codesome): persist uncommittedRec with checkpoint's metadata
-		err := b.kafkaClient.CommitRecords(ctx, checkpointRec)
-		if err != nil {
-			// TODO(codesome): add metric
-			return err
+	ctx = kgo.PreCommitFnContext(ctx, func(req *kmsg.OffsetCommitRequest) error {
+		meta := marshallCommitMeta(commitRec.Timestamp.UnixMilli(), lastRec.Timestamp.UnixMilli(), blockEnd)
+		for _, topic := range req.Topics {
+			if topic.Topic != b.cfg.Kafka.Topic {
+				continue
+			}
+			for i := range req.Topics[0].Partitions {
+				if req.Topics[0].Partitions[i].Partition == commitRec.Partition {
+					req.Topics[0].Partitions[i].Metadata = &meta
```

Comment:
Fixed the commit metadata. Debugging through the kgo library I saw that this string was not being overridden. By default kgo sets it to member ID (which is what we were seeing).

See https://github.com/grafana/mimir/pull/8329/commits/3e5e03f3f7bf0ecfe4019fab85e1d1b1fa02bd3d for the fix